File: problems/0886_possible-bipartition/solution_dfs_color.ts
function possibleBipartition(n: number, dislikes: number[][]): boolean {
    // Dislike is symmetric, so build an undirected adjacency list: a
    // valid two-group split is exactly a 2-coloring of this graph.
    const adjacency: number[][] = Array.from({ length: n + 1 }, () => []);
    for (const [a, b] of dislikes) {
        adjacency[a].push(b);
        adjacency[b].push(a);
    }

    const color: number[] = new Array(n + 1).fill(0); // 0 = uncolored, 1 / -1 = the two groups
    // The dislike graph may be disconnected, so the scan restarts the
    // DFS from every still-uncolored person; each run colors one
    // whole connected component.
    for (let start = 1; start <= n; start++) {
        if (color[start] !== 0) {
            continue;
        }
        color[start] = 1;
        // The stack drives a depth-first sweep: pop a person, then
        // push every uncolored neighbor with the opposite color
        // (marking on push); a neighbor already sharing the current
        // color closes an odd cycle, so no split exists.
        const stack: number[] = [start];
        while (stack.length > 0) {
            const person = stack.pop()!;
            for (const neighbor of adjacency[person]) {
                if (color[neighbor] === 0) {
                    color[neighbor] = -color[person];
                    stack.push(neighbor);
                } else if (color[neighbor] === color[person]) {
                    return false;
                }
            }
        }
    }
    return true;
}

File: problems-adapt/0801-0900/0886_two-conflict-free-groups/solution_dfs_color.py
class Solution:
    def canSplitInTwo(self, n: int, conflicts: list[list[int]]) -> bool:
        # A conflict runs both ways, so build an undirected adjacency list: a
        # valid two-group split is exactly a 2-coloring of this graph.
        adjacency = [[] for _ in range(n + 1)]
        for a, b in conflicts:
            adjacency[a].append(b)
            adjacency[b].append(a)

        color = [0] * (n + 1)  # 0 = uncolored, 1 / -1 = the two groups
        # The conflict graph may be disconnected, so the scan restarts the
        # DFS from every still-uncolored person; each run colors one
        # whole connected component.
        for start in range(1, n + 1):
            if color[start] != 0:
                continue
            color[start] = 1
            # The stack drives a depth-first sweep: pop a person, then
            # push every uncolored neighbor with the opposite color
            # (marking on push); a neighbor already sharing the current
            # color closes an odd cycle, so no split exists.
            stack = [start]
            while stack:
                person = stack.pop()
                for neighbor in adjacency[person]:
                    if color[neighbor] == 0:
                        color[neighbor] = -color[person]
                        stack.append(neighbor)
                    elif color[neighbor] == color[person]:
                        return False
        return True
